# Conhecendo o unittest

> ðŸŽ¬ Caso vocÃª prefira consumir este conteÃºdo em vÃ­deo, ele estÃ¡ disponÃ­vel no final do tÃ³pico. Lembre-se de que os cÃ³digos apresentados no vÃ­deo foram construÃ­dos ao longo do conteÃºdo escrito e que reproduzir os passos demonstrados durante o estudo Ã© importante para o aprendizado. ðŸ˜‰

OÂ `unittest`Â Ã© um mÃ³dulo de teste em Python que faz parte da biblioteca padrÃ£o. Ele fornece uma estrutura para escrever e executar testes automatizados, a fim de verificar a correÃ§Ã£o do cÃ³digo. OÂ `unittest`Â permite que vocÃª defina testes unitÃ¡rios, que sÃ£o testes focados em verificar a funcionalidade individual de partes isoladas do seu cÃ³digo.

Com oÂ `unittest`, vocÃª pode criar classes de teste que herdam da classeÂ `unittest.TestCase`. Dentro dessas classes de teste, vocÃª pode definir mÃ©todos de teste, que sÃ£o funÃ§Ãµes que verificam se um determinado comportamento do cÃ³digo estÃ¡ de acordo com o esperado. Esses mÃ©todos de teste podem conter chamadas a vÃ¡rias asserÃ§Ãµes fornecidas peloÂ `unittest`, que verificam se determinadas condiÃ§Ãµes sÃ£o verdadeiras.

AlÃ©m disso, oÂ `unittest`Â tambÃ©m fornece recursos para configurar e limpar o ambiente de teste, lidar com exceÃ§Ãµes e erros esperados, e executar testes em vÃ¡rios contextos.

## Como criar testes com o unittest

OÂ `unittest`Â utiliza quatro conceitos cruciais para construir a estrutura de testes:Â `test fixture`Â (ambiente de teste),Â `test case`Â (casos de teste),Â `test suite`Â (conjuntos de teste) eÂ `test runner`Â (executor de teste).

- `Test fixture`Â - refere-se a funÃ§Ãµes que sÃ£o executadas antes e/ou depois dos testes, geralmente criando um contexto necessÃ¡rio para a execuÃ§Ã£o dos testes ou realizando tarefas de limpeza apÃ³s a conclusÃ£o dos mesmos. Ã‰ comum utilizar fixtures para criar dados que serÃ£o utilizados por vÃ¡rios testes.

- `Test case`Â - representa um caso de teste especÃ­fico que verifica uma resposta particular para uma determinada execuÃ§Ã£o de uma funÃ§Ã£o ou fluxo.

- `Test suite`Â - consiste em coleÃ§Ãµes de casos de teste, podendo incluir outras suites de teste ou uma combinaÃ§Ã£o de ambas. As suites sÃ£o utilizadas para agrupar testes que devem ser executados em conjunto ou que possuem uma relaÃ§Ã£o lÃ³gica entre si.

- `Test runner`Â - Ã© um componente responsÃ¡vel por orquestrar a execuÃ§Ã£o dos testes e fornecer os resultados ao usuÃ¡rio. O executor pode utilizar uma interface grÃ¡fica, uma interface textual ou retornar um valor especial para indicar os resultados da execuÃ§Ã£o dos testes.

Vamos a um exemplo, digamos que nÃ³s temos uma funÃ§Ã£o bÃ¡sica de divisÃ£o, passamos 2 nÃºmeros como argumentos para essa funÃ§Ã£o e ela divide o primeiro pelo segundo:

> divisao.py

Copiar

```python
def divide(a, b):
    return a / b
```

Vamos testar a funcionalidade dessa aplicaÃ§Ã£o a partir noÂ `unittest`:

> test_divisao.py

Copiar

```python
import unittest

from divisao import divide


class TestOperations(unittest.TestCase):
    def test_divide(self):
        self.assertEqual(divide(5, 2), 2.5)


if __name__ == '__main__':
    unittest.main()
```

Primeiramente importamos o moduloÂ `unittest`Â e, em seguida, importamos a funÃ§Ã£oÂ `divide`Â do mÃ³duloÂ `divisao`. Em seguida, criamos a classeÂ `TestOperations`Â (que poderia ter qualquer nome) que herda da classeÂ `unittest.TestCase`. Essa heranÃ§a permite que a classeÂ `TestOperations`Â tenha acesso a todos os mÃ©todos e propriedades fornecidos peloÂ `unittest.TestCase`.

Dentro da classeÂ `TestOperations`, definimos o mÃ©todoÂ `test_divide`, que recebe o parÃ¢metroÂ `self`. Esse parÃ¢metro permite o acesso aos mÃ©todos de asserÃ§Ã£o fornecidos peloÂ `unittest`Â por meio da heranÃ§a de classe. Nesse caso, utilizamos o mÃ©todoÂ `assertEqual`, que verifica se o primeiro parÃ¢metro (`divide(5, 2)`) Ã© igual ao segundo parÃ¢metro (`2.5`), indicando assim se a funÃ§Ã£oÂ `divide(5, 2)`Â retorna o resultado esperado.

Para executar esse teste, basta rodar o seguinte comando:

Copiar

```bash
python3 -m unittest test_divisao.py
```

SaÃ­da esperada:

|![teste divide sem detalhes sucesso](https://content-assets.betrybe.com/prod/7d17a303-4111-4f9c-a460-1f42bd0054fd-teste%20divide%20sem%20detalhes%20sucesso.png)|
|---|
|Teste sem detalhes|

VocÃª pode rodar os testes com mais detalhes passando a flagÂ `-v`:

Copiar

```bash
python3 -m unittest -v test_divisao.py
```

SaÃ­da esperada:

|![teste divide com detalhes sucesso](https://content-assets.betrybe.com/prod/7d17a303-4111-4f9c-a460-1f42bd0054fd-teste%20divide%20com%20detalhes%20sucesso.png)|
|---|
|Teste com detalhes|

Existem outras assertions que podemos utilizar nos nossos testes, dentre elas existe aÂ `assertRaises()`Â que espera que uma exceÃ§Ã£o seja lanÃ§ada.

> test_divisao.py

Copiar

```python
# ...
    # def test_divide(self):
    #     self.assertEqual(divide(5, 2), 2.5)

    def test_divide_by_0(self):
        with self.assertRaises(ZeroDivisionError):
            divide(5, 0)
# ...
```

Copiar

```bash
python3 -m unittest -v test_divisao.py
```

SaÃ­da esperada:

|![teste divide com 0](https://content-assets.betrybe.com/prod/7d17a303-4111-4f9c-a460-1f42bd0054fd-teste%20divide%20com%200.png)|
|---|
|Teste divisÃ£o por 0 com detalhes|

Quando tentamos dividir um nÃºmero por 0, o Python lanÃ§a a exceÃ§Ã£oÂ `ZeroDivisionError`, usando oÂ `assertRaises`Â garantimos que o lanÃ§amento de uma exceÃ§Ã£o serÃ¡ testado da forma correta.

Como podemos ver, nossos testes funcionaram da forma certa, mas o que acontece se algum caso teste der errado? Vamos alterar o resultado esperado de um dos nossos testes para ver o que acontece:

> test_divisao.py

Copiar

```python
# ...
    # def test_divide(self):
        self.assertEqual(divide(5, 2), 2.4)
# ...
```

Copiar

```bash
python3 -m unittest -v test_divisao.py
```

SaÃ­da esperada:

|![teste divide com detalhes falha](https://content-assets.betrybe.com/prod/7d17a303-4111-4f9c-a460-1f42bd0054fd-teste%20divide%20com%20detalhes%20falha.png)|
|---|
|Teste de falha com detalhes|

Na execuÃ§Ã£o do testeÂ `test_divide`, ocorre umÂ `AssertionError`. No entanto, esse erro nÃ£o impede a execuÃ§Ã£o do testeÂ `test_divide_by_0`. Os erros em um caso de teste nÃ£o interferem nos outros casos de teste. Em um cenÃ¡rio de testes, quando um teste falha e lanÃ§a uma exceÃ§Ã£o, nÃ£o Ã© necessÃ¡rio tratÃ¡-la comÂ `try/except`. Em vez disso, o erro deve ser corrigido na implementaÃ§Ã£o do teste ou no cÃ³digo sendo testado. No nosso caso, o teste em si estÃ¡ com um problema, nÃ£o o cÃ³digo sendo testado. Voltando o resultado esperado paraÂ `2.5`Â o teste deve funcionar da forma esperada.

## Testando Flask com unittest

Agora que vocÃª jÃ¡ sabe os fundamentos de testes com oÂ `unittest`, vamos aplicar esse conhecimento em uma aplicaÃ§Ã£o Flask. Para isso, vamos utilizar uma aplicaÃ§Ã£o de exemplo que estÃ¡ disponÃ­vel no seguinte repositÃ³rio:Â [tryber/python-conteudo-flask-project-manager](https://github.com/tryber/python-conteudo-flask-project-manager). FaÃ§a o clone desse repositÃ³rio e siga as instruÃ§Ãµes noÂ `README.md`Â para deixar a aplicaÃ§Ã£o funcional!

A aplicaÃ§Ã£o deste repositÃ³rio mostra tarefas de alguns projetos, atualmente a aplicaÃ§Ã£o sÃ³ mostra as tarefas e projetos jÃ¡ cadastrados no nosso banco de dados. Sinta-se a vontade para criar novas features e usar os conhecimentos de hoje para testÃ¡-las! ðŸ˜‰

NÃ³s vamos utilizar oÂ `unittest`Â para testar o retorno da nossa requisiÃ§Ã£o, ou seja, vamos testar o que estÃ¡ sendo renderizado noÂ _front-end_.

Vamos comeÃ§ar pela tela inicial, a telaÂ `home`, que pode ser acessada pelas rotasÂ `/`Â eÂ `/projects`, vamos comeÃ§ar fazendo a preparaÃ§Ã£o para os testes. Vamos criar um arquivo chamadoÂ `test_home_unittest.py`Â e adicionar o seguinte cÃ³digo a ele:

Copiar

```bash
mkdir src/tests && touch src/tests/test_home_unittest.py
```

> src/tests/test_home_unittest.py

Copiar

```python
import unittest
from app import app


class TestHome(unittest.TestCase):
    def setUp(self):
        test_app = app.test_client()
        self.response = test_app.get('/')
```

No cÃ³digo acima definimos a classeÂ `TestHome`, que herda da classeÂ `unittest.TestCase`, como vimos anteriormente.

Dentro da classeÂ `TestHome`, hÃ¡ um mÃ©todo chamadoÂ `setUp()`. Esse mÃ©todo Ã© executado antes de cada caso de teste e Ã© usado para configurar o ambiente de teste. Dentro do mÃ©todoÂ `setUp()`, criamos uma instÃ¢ncia deÂ `app.test_client()`, essa instÃ¢ncia nos permite simular requisiÃ§Ãµes Ã  aplicaÃ§Ã£o. Em seguida, fazemos uma solicitaÃ§Ã£oÂ `GET`Â Ã  rota principal (`/`) da aplicaÃ§Ã£o atravÃ©s da instÃ¢ncia deÂ `app.test_client()`, armazenando a resposta retornada emÂ `self.response`. Essa linha simula o acesso Ã  pÃ¡gina inicial da aplicaÃ§Ã£o e captura a resposta retornada pelo servidor.

Vamos criar nosso primeiro teste e verificar o status da resposta da nossa requisiÃ§Ã£o:

> src/tests/test_home_unittest.py

Copiar

```python
# ...
    # def setUp(self):
    #     test_app = app.test_client()
    #     self.response = test_app.get('/')

    def test_status_response(self):
        self.assertEqual(self.response.status_code, 200)
```

Execute os testes:

Copiar

```bash
docker compose exec flask-api python3 -m unittest -v tests/test_home_unittest.py -b
```

SaÃ­da esperada:

|![Test status response](https://content-assets.betrybe.com/prod/997505fc-d4ce-46fc-a3c5-ed9a1cb59193-Test%20status%20response.png)|
|---|
|Test status response|

Vamos verificar se foram criados containers para cada um dos nossos 3 projetos, para isso vamos criar um diretÃ³rio chamadoÂ `mocks`Â e um arquivo chamadoÂ `home.py`:

Copiar

```bash
mkdir src/tests/mocks && touch src/tests/mocks/home.py
```

> src/tests/mocks/home.py

Copiar

```python
project_cards = '<section class="project-home">'
```

> src/tests/test_home_unittest.py

Copiar

```python
# import unittest
# from app import app
from tests.mocks.home import project_cards
# ...
    # def test_status_response(self):
    #     self.assertEqual(self.response.status_code, 200)

    def test_quantity_of_projects(self):
        self.assertEqual(self.response.text.count(project_cards), 3)
```

Execute os testes:

Copiar

```bash
docker compose exec flask-api python3 -m unittest -v tests/test_home_unittest.py -b
```

SaÃ­da esperada:

|![Test quantity of projects](https://content-assets.betrybe.com/prod/997505fc-d4ce-46fc-a3c5-ed9a1cb59193-Test%20quantity%20of%20projects.png)|
|---|
|Test quantity of projects|

Vamos agora testar se os tÃ­tulos dos 3 projetos estÃ£o sendo renderizados, criaremos um dicionÃ¡rio para guardar as informaÃ§Ãµes dos tÃ­tulos dos 3 projetos que queremos testar:

> src/mocks/home.py

Copiar

```python
# ...

project_titles = {
    1: "<h1>AplicaÃ§Ã£o Web para Gerenciamento de Tarefas</h1>",
    2: "<h1>Aplicativo mÃ³vel para rastreamento de exercÃ­cios fÃ­sicos</h1>",
    3: "<h1>Sistema de gestÃ£o de vendas online</h1>"
}
```

> src/tests/test_home_unittest.py

Copiar

```python
# ...
from tests.mocks.home import project_cards, project_titles
# ...
    # def test_status_response(self):
    #     self.assertEqual(self.response.status_code, 200)

    def test_projects_titles(self):
        self.assertTrue(project_titles[1] in self.response.text)
        self.assertTrue(project_titles[2] in self.response.text)
        self.assertTrue(project_titles[3] in self.response.text)
```

Execute os testes:

Copiar

```bash
docker compose exec flask-api python3 -m unittest -v tests/test_home_unittest.py -b
```

SaÃ­da esperada:

![Test projects titles](https://content-assets.betrybe.com/prod/997505fc-d4ce-46fc-a3c5-ed9a1cb59193-Test%20projects%20titles.png)|:â€“:| |Test projects titles|

Vamos testar uma rota desconhecida e qual serÃ¡ o seu retorno:

> src/tests/mocks/home.py

Copiar

```python
# ...

unknown_page = {
    'title': '<h1>Not Found</h1>',
    'text': '<p>The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.</p>'
}
```

> src/tests/test_home_unittest.py

Copiar

```python
# ...
from tests.mocks.home import project_cards, project_titles, unknown_page
# ...
    # def test_projects_titles(self):
    #     self.assertTrue(project_titles[1] in self.response.text)
    #     self.assertTrue(project_titles[2] in self.response.text)
    #     self.assertTrue(project_titles[3] in self.response.text)

    def test_unknown_route(self):
        response = app.test_client().get("/unknown")
        self.assertEqual(response.status_code, 404)
        self.assertTrue(unknown_page['title'] in response.text)
        self.assertTrue(unknown_page['text'] in response.text)
```

Execute os testes:

Copiar

```bash
docker compose exec flask-api python3 -m unittest -v tests/test_home_unittest.py -b
```

SaÃ­da esperada:

|![Test unknown route](https://content-assets.betrybe.com/prod/997505fc-d4ce-46fc-a3c5-ed9a1cb59193-Test%20unknown%20route.png)|
|---|
